### ISSUE TITLE
How to preserve zoom settings across callbacks/page refreshes?

### ISSUE THREAD

**Issue by perfectly-preserved-pie** [thedirtyfew/dash-leaflet#144]
Hi,

I'm basically wondering how to implement this: https://iotespresso.com/preserving-zoom-settings-across-data-refresh-in-dash/
in Dash Leaflet.

I have a bunch of checkboxes that filter through a dataframe, and a map that plots points based on that filtered dataframe. Whenever I check or uncheck a checkbox, the zoom level gets reset and it's very jarring/annoying. I'm hoping I can just have the zoom level stay where it is when a checkbox is un/checked.

**Comment by emilhe** [comment]
Could you provide a minimal, runnable example demonstrating the issue?

**Comment by perfectly-preserved-pie** [comment]
Sure, here it is

```
# import the csv
# Don't round the float. See https://stackoverflow.com/a/68027847
df = pd.read_csv("finallll csv lol.csv", float_precision="round_trip")
pd.set_option("display.precision", 10)

# Define HTML code for the popup so it looks pretty and nice
def popup_html(row):
    i = row.Index
    street_address=df['Full Street Address'].iloc[i] 
    mls_number=df['Listing ID (MLS#)'].iloc[i]
    mls_number_hyperlink=f"https://www.bhhscalifornia.com/for-lease/{mls_number}-t_q;/"
    lc_price = df['L/C Price'].iloc[i] 
    price_per_sqft=df['Price Per Square Foot'].iloc[i]                  
    brba = df['Br/Ba'].iloc[i]
    square_ft = df['Sqft'].iloc[i]
    year = df['YrBuilt'].iloc[i]
    garage = df['Garage Spaces'].iloc[i]
    pets = df['PetsAllowed'].iloc[i]
    phone = df['List Office Phone'].iloc[i]
    terms = df['Terms'].iloc[i]
    sub_type = df['Sub Type'].iloc[i]
    # Return the HTML snippet but NOT as a string. See https://github.com/thedirtyfew/dash-leaflet/issues/142#issuecomment-1157890463 
    return [
      html.Table([ # Create the table
        html.Tbody([ # Create the table body
          html.Tr([ # Start row #1
            html.Td("Street Address"), html.Td(f"{street_address}")
          ]), # end row #1
          html.Tr([ # Start row #2
            # Use a hyperlink to link to BHHS, don't use a referrer, and open the link in a new tab
            # https://www.freecodecamp.org/news/how-to-use-html-to-open-link-in-new-tab/
            html.Td("Listing ID (MLS#)"), html.Td(html.A(f"{mls_number}", href=f"{mls_number_hyperlink}", referrerPolicy='noreferrer', target='_blank'))
          ]), # end row #2
          html.Tr([ # Start row #3
            html.Td("L/C Price"), html.Td(f"${lc_price}")
          ]), # end row #3
          html.Tr([
            html.Td("Price Per Square Foot"), html.Td(f"{price_per_sqft}")
          ]),
          html.Tr([
            html.Td("Bedrooms/Bathrooms"), html.Td(f"{brba}")
          ]),
          html.Tr([
            html.Td("Square Feet"), html.Td(f"{square_ft}")
          ]),
          html.Tr([
            html.Td("Year Built"), html.Td(f"{year}")
          ]),
          html.Tr([
            html.Td("Garage Spaces"), html.Td(f"{garage}"),
          ]),
          html.Tr([
            html.Td("Pets Allowed?"), html.Td(f"{pets}"),
          ]),
          html.Tr([
            html.Td("List Office Phone"), html.Td(f"{phone}"),
          ]),
          html.Tr([
            html.Td("Rental Terms"), html.Td(f"{terms}"),
          ]),
          html.Tr([                                                                                            
            html.Td("Physical Sub Type"), html.Td(f"{sub_type}")                                                                                    
          ]), # end rows
        ]), # end body
      ]), # end table
    ]

# Create markers & associated popups from dataframe
markers = [dl.Marker(children=dl.Popup(popup_html(row)), position=[row.Latitude, row.Longitude]) for row in df.itertuples()]

# Get the means so we can center the map
lat_mean = df['Latitude'].mean()
long_mean = df['Longitude'].mean()

# Add them to a MarkerCluster
cluster = dl.MarkerClusterGroup(id="markers", children=markers)

app = JupyterDash(__name__, external_stylesheets=external_stylesheets)

app.layout = html.Div([
  # Create a checklist of options for the user
  # https://dash.plotly.com/dash-core-components/checklist
  dcc.Checklist( 
      id = 'subtype_checklist',
      options=[
        {'label': 'Apartment (Attached)', 'value': 'APT/A'},
        {'label': 'Studio (Attached)', 'value': 'STUD/A'},
        {'label': 'Single Family Residence (Attached)', 'value': 'SFR/A'},
        {'label': 'Single Family Residence (Detached)', 'value': 'SFR/D'},
        {'label': 'Condo (Attached)', 'value': 'CONDO/A)'},
        {'label': 'Condo (Detached)', 'value': 'CONDO/D'},
        {'label': 'Quadplex (Attached)', 'value': 'QUAD/A'},
        {'label': 'Quadplex (Detached)', 'value': 'QUAD/D'},
        {'label': 'Triplex (Attached)', 'value': 'TPLX/A'},
        {'label': 'Townhouse (Attached)', 'value': 'TWNHS/A'},
        {'label': 'Townhouse (Detached)', 'value': 'TWNHS/D'},
        {'label': 'Duplex (Attached)', 'value': 'DPLX/A'},
        {'label': 'Duplex (Detached)', 'value': 'DPLX/D'},
        {'label': 'Ranch House (Detached)', 'value': 'RMRT/D'}
      ],
      value=['APT/A'] # Set the default value
  ),
  # Create a checklist for pet policy
  dcc.Checklist(
    id = 'pets_checklist',
    options=[
      {'label': 'Pets Allowed', 'value': 'True'},
      {'label': 'Pets NOT Allowed', 'value': 'False'}
    ],
      value=['True', 'False'] # A value needs to be selected upon page load otherwise we error out. See https://community.plotly.com/t/how-to-convert-a-nonetype-object-i-get-from-a-checklist-to-a-list-or-int32/26256/2
  ),
  # Create a checklist for rental terms
  dcc.Checklist(
    id = 'terms_checklist',
    options = [
      {'label': 'Monthly', 'value': 'MO'},
      {'label': '12 Months', 'value': '12M'},
      {'label': '24 Months', 'value': '24M'},
      {'label': 'Negotiable', 'value': 'NG'}
    ],
      value=['MO', '12M', '24M', 'NG']
  ),
  # Create a range slider for # of garage spaces
  dcc.RangeSlider(
    min=0, 
    max=df['Garage Spaces'].max(), # Dynamically calculate the maximum number of garage spaces
    step=1, 
    value=[0, df['Garage Spaces'].max()], 
    id='garage_spaces_slider'
  ),
  # Create a range slider for rental price
  dcc.RangeSlider(
    min=df['L/C Price'].min(),
    max=df['L/C Price'].max(),
    value=[0, df['L/C Price'].max()],
    id='rental_price_slider',
    tooltip={
      "placement": "bottom",
      "always_visible": True
    }
  ),

  # Generate the map
  dl.Map(
    [dl.TileLayer(), cluster],
    id='map',
    zoom=3,
    center=(lat_mean, long_mean),
    style={'width': '100%', 'height': '50vh', 'margin': "auto", "display": "block"}
  )

])

@app.callback(
  Output(component_id='map', component_property='children'),
  [
    Input(component_id='subtype_checklist', component_property='value'),
    Input(component_id='pets_checklist', component_property='value'),
    Input(component_id='terms_checklist', component_property='value'),
    Input(component_id='garage_spaces_slider', component_property='value'),
    Input(component_id='rental_price_slider', component_property='value')
  ]
)
def update_map(subtypes_chosen, pets_chosen, terms_chosen, garage_spaces, rental_price):
  df_filtered = df[
    (
      (df['Sub Type'].isin(subtypes_chosen)) &
      (df['PetsAllowedSimple'].isin(pets_chosen)) &
      (df['Terms'].isin(terms_chosen))
    ) &
    # For the slider, we need to filter the dataframe by an integer range this time and not a string like the ones aboves
    # To do this, we can use the Pandas .between function
    # See https://stackoverflow.com/a/40442778
    (df['Garage Spaces'].between(garage_spaces[0], garage_spaces[1])) &
    # Repeat but for rental price
    (df['L/C Price'].between(rental_price[0], rental_price[1]))
  ]

  # Create markers & associated popups from dataframe
  markers = [dl.Marker(children=dl.Popup(popup_html(row)), position=[row.Latitude, row.Longitude]) for row in df_filtered.itertuples()]

  # Add them to a MarkerCluster
  cluster = dl.MarkerClusterGroup(id="markers", children=markers)

  # Get the means so we can center the map
  lat_mean = df['Latitude'].mean()
  long_mean = df['Longitude'].mean()

  # Generate the map
  return dl.Map(
    [dl.TileLayer(), cluster],
    id='map',
    zoom=3,
    center=(lat_mean, long_mean),
    style={'width': '100%', 'height': '50vh', 'margin': "auto", "display": "block"}
  )



# Launch the Flask app
if __name__ == '__main__':
    app.run_server(mode='external', host='192.168.4.196', port='9208', debug='false')
```

CSV attached
[finallll csv lol.csv](https://github.com/thedirtyfew/dash-leaflet/files/8934345/finallll.csv.lol.csv)

**Comment by perfectly-preserved-pie** [comment]
Also, here's a GIF of the behavior I'm seeing:

https://i.imgur.com/lhI4UCL.mp4

**Comment by emilhe** [comment]
The issue is due to the fact that you are replacing the whole `Map` element. Instead, I would recommend to wrap the `cluster` is a `LayerGroup`, 

```
dl.Map(
    [dl.TileLayer(), dl.LayerGroup(id="cluster")],
    id='map',
    zoom=3,
    center=(lat_mean, long_mean),
    style={'width': '100%', 'height': '50vh', 'margin': "auto", "display": "block"}
  )
```

and then replace it's children,

```
@app.callback(
  Output(component_id='cluster', component_property='children'), ...
) 
...
return dl.MarkerClusterGroup(id=str(uuid.uuid4()), children=markers)
```

**Comment by perfectly-preserved-pie** [comment]
Hmm... that makes sense. I thought it was weird that I was regenerating the entire map. But when I make those changes, the resulting map doesn't have any marked points _at all_. The zoom thing has been fixed though.

This is my dl.Map code right above the callback:

```
  # Generate the map
  dl.Map(
    [dl.TileLayer(), dl.LayerGroup(id='cluster')],
    id='map',
    zoom=3,
    center=(lat_mean, long_mean),
    style={'width': '100%', 'height': '50vh', 'margin': "auto", "display": "block"}
  )
```

and the actual modified callback:
```
@app.callback(
  Output(component_id='cluster', component_property='children'),
  [
    Input(component_id='subtype_checklist', component_property='value'),
    Input(component_id='pets_checklist', component_property='value'),
    Input(component_id='terms_checklist', component_property='value'),
    Input(component_id='garage_spaces_slider', component_property='value'),
    Input(component_id='rental_price_slider', component_property='value'),
    Input(component_id='bedrooms_slider', component_property='value'),
    Input(component_id='bathrooms_slider', component_property='value'),
  ]
)
def update_map(subtypes_chosen, pets_chosen, terms_chosen, garage_spaces, rental_price, bedrooms_chosen, bathrooms_chosen):
  df_filtered = df[
    (df['Sub Type'].isin(subtypes_chosen)) &
    (df['PetsAllowedSimple'].isin(pets_chosen)) &
    (df['Terms'].isin(terms_chosen)) &
    # For the slider, we need to filter the dataframe by an integer range this time and not a string like the ones aboves
    # To do this, we can use the Pandas .between function
    # See https://stackoverflow.com/a/40442778
    (df['Garage Spaces'].between(garage_spaces[0], garage_spaces[1])) &
    # Repeat but for rental price
    (df['L/C Price'].between(rental_price[0], rental_price[1])) &
    (df['Bedrooms'].between(bedrooms_chosen[0], bedrooms_chosen[1])) &
    (df['Total Bathrooms'].between(bathrooms_chosen[0], bathrooms_chosen[1]))
  ]

  # Create markers & associated popups from dataframe
  markers = [dl.Marker(children=dl.Popup(popup_html(row)), position=[row.Latitude, row.Longitude]) for row in df_filtered.itertuples()]

  # Add them to a MarkerCluster
  cluster = dl.MarkerClusterGroup(id="markers", children=markers)

  # Get the means so we can center the map
  lat_mean = df['Latitude'].mean()
  long_mean = df['Longitude'].mean()

  return dl.MarkerClusterGroup(
    id='cluster',
    children=markers
  )

# Launch the Flask app
if __name__ == '__main__':
    app.run_server(mode='external', host='192.168.4.196', port='9208', debug='false')
```

What did I miss?

**Comment by emilhe** [comment]
It might be because the ID of the marker cluster is fixed, could you try making it variable as in my pseudo code?

**Comment by perfectly-preserved-pie** [comment]
Even with 

```
return dl.MarkerClusterGroup(
    id=str(uuid.uuid4()),
    children=markers
  )
```
the map is still blank. Why would it need to be variable though?

**Comment by perfectly-preserved-pie** [comment]
Do you have ideas? I've tried messing with the MarkerClusterGroup id every which way, but the markers still don't appear on the map.

**Comment by perfectly-preserved-pie** [comment]
Ok, I apologize. The fix was a simple restart. 🤦‍♂️ I got desperate and I tried going back to my "Last Known Working Condition" of my code just to start from a fresh base, and it was _still_ broken as in no markers were being placed on the map. I decided to restart my Python kernel to go one step further, and after restarting it your code modifications worked perfectly and everything is working correctly.

I guess tinkering in Jupyter for weeks straight caused some pollution of variables or something that was interfering with my most recent tries. Only a kernel restart could clear it up.

Thank you very much for your assistance; you've helped solve a major roadblock in my project. I appreciate it.
